Add coercing in JSONExtract

marcioapm opened the pull request:
### Changelog category (leave one):
- Backward Incompatible Change

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
JSONExtract family of functions will now attempt to coerce to the request type.


JSONExtract family of functions do not support (U)Int128+, and is currently too type-strict, making it very hard to deal with JSON produced by different encoders, or that we simply don't control.

This PR attempts to fix some of these limitations, and adds support for (U)Int128, and (U)Int256.
I also noticed some issues with parsing of these non-native types, and will likely fix them in another PR.

marcioapm committed:
Add Element::type() to JSONParser

marcioapm committed:
Add convenience ctor to ReadBufferFromMemory

marcioapm committed:
Remove redundant default case in switch statement

marcioapm committed:
Allow coercing in JSONExtract* functions

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=41502) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=41502) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=41502) it.</sub>

marcioapm commented:
@Avogar so the two tests that fail are basically due to strictness of the current implementation, for example, converting `Float64` `1.1` to `Decimal(0)` will fail and return `0` with the current implementation, but will be converted to `Decimal(0)` `1` with the changes in my PR.
I made it so that parsing and converting to decimal is a little less strict. Currently it can only be exact, 0 or null.
I can either change the test so that `toDecimal` becomes less strict, my preferred option, or add a template parameter and use the less strict parsing only on the JSONExtract* functions. From my experience, I prefer it less strict, but am always reluctant to change current behavior.
What do you think would be a better approach?

Avogar commented:
Sory for long response, let's change existing test, I think new behaviour will be better (but need to think about adding special setting that will allow to get back to previous behaviour, not sure it it's really neeed)

marcioapm commented:
@Avogar So the Fuzzer test fail does not seem related to my change, but it is perhaps worth fixing.
Essentially, in src/Functions/FunctionsJSON.cpp:194 when validating the parameters, it's checking only if it's an integer, but then when actually using this parameter, it expects it to be a Int64, so it crashes, because the fuzzer passed it a UInt128.

This the check:
```c++
            if (!isString(column.type) && !isInteger(column.type))
                throw Exception{"The argument " + std::to_string(i + 1) + " of function " + String(function_name)
                                    + " should be a string specifying key or an integer specifying index, illegal type: " + column.type->getName(),
                                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT};
```

This is where it crashes:
```c++
                case MoveType::Index:
                {
                    Int64 index = (*arguments[j + 1].column)[row].get<Int64>();
                    if (!moveToElementByIndex<JSONParser>(res_element, index, key))
                        return false;
                    break;
                }
```

I think the correct fix would be to change the parameter check from `isInteger` to `isUInt64`.
Shall I make a new PR with a fix, or can I just fix it on this PR?

Performance test is also failing but completely unrelated, I am no sure what I should do next to get this merged.

marcioapm commented:
@Avogar ping

Avogar reviewed (approved):
Great work, thank you! Just two small comments. 

Sorry for long review, had a vacation and a lot of work recently.

marcioapm commented on src/IO/readDecimalText.h:
```diff
@@ -147,23 +147,30 @@ inline bool readDigits(ReadBuffer & buf, T & x, uint32_t & digits, int32_t & exp
 
-template <typename T>
-inline void readDecimalText(ReadBuffer & buf, T & x, uint32_t precision, uint32_t & scale, bool digits_only = false)
+template <typename T, bool _throw_on_error=true>
```
It his a pattern used often in the codebase? In my opinion, using the return type template argument to control the throw behaviour is not explicit/clear enough for me. But if it's used often in the codebase, than I suppose that would be fine.

Avogar replied:
Yes, we use this pattern often (you can see examples in ReadHelpers)

marcioapm replied:
Alright, I will update it.

marcioapm commented on src/Functions/FunctionsJSON.cpp:
```diff
@@ -623,24 +623,32 @@ class JSONTypeImpl
-            type = 0;
-        else
-            return false;
+        switch (element.type())
+        {
+            case ElementType::INT64:
+                type = 'i';
+                break;
+            case ElementType::UINT64:
+                type = 'u';
+                break;
+            case ElementType::DOUBLE:
+                type = 'd';
+                break;
+            case ElementType::STRING:
+                type = '"';
+                break;
+            case ElementType::ARRAY:
+                type = '[';
+                break;
+            case ElementType::OBJECT:
+                type = '{';
+                break;
+            case ElementType::NULL_VALUE:
+                type = 0;
+                break;
```
The return of `type()` is already mapped to the same values of the `simdjson` implementation, so I suspect the compiler will replace any switches on this with just the result of the type() call, if it was inlined, at least for the simdjson specialization, but I haven't verified it.
I would like to preserve that, in the general case. Here, most enum member values also match, except for null (`0` vs `n`), but I would still keep this as it, because this is the implementation of `JSONType`, but the call to `type()` happens in a lot of other places, in hotter code paths as well.

marcioapm committed:
Use return type template parameter to select throwing behavior

marcioapm commented:
@Avogar anything else I need to do?

Avogar commented:
ASTFuzzer failure is related to changes: `SELECT JSONExtract('{"a":1234567890123456789}', 'a', JSONExtract('aa', 'UInt256'), 'UInt256') AS a, toTypeName(a)` causes logical error `Invalid Field get from type UInt256 to type Int64`

marcioapm commented:
> ASTFuzzer failure is related to changes: `SELECT JSONExtract('{"a":1234567890123456789}', 'a', JSONExtract('aa', 'UInt256'), 'UInt256') AS a, toTypeName(a)` causes logical error `Invalid Field get from type UInt256 to type Int64`

Yes, please see my earlier message about the Fuzzer, and let me know how to proceed.

marcioapm committed:
Fix array index argument type check for JSONExtract* functions

marcioapm committed:
Merge branch 'master' into coercing-json-extract

marcioapm commented:
> Oh, sorry, missed that message. We should replace `isInteger` with `isNativeInteger` that doesn't check `(U)Int128/(U)Int256`

Fixed with last commit. Also added a test.
